doc/man/pilot-nredir.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "pilot-nredir" 1 "Copyright 1996-2007 FSF" "0.12.5" "PILOT-LINK"
.SH NAME
pilot-nredir  \- Accept connection and redirect via Network Hotsync Protocol.
.SH "SECTION"

.PP
pilot\-link: Tools

.SH "SYNOPSIS"

.PP
 \fIpilot\-nredir\fR [\fB\-p\fR|\fB\-\-port\fR <\fBport\fR>] [\fB\-\-version\fR] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-?\fR|\fB\-\-help\fR] [\fB\-\-usage\fR] [\fB\-n\fR|\fB\-\-net\fR]

.SH "DESCRIPTION"

.PP
This will bind your locally connected device to a network port, and redirect them through the network device to a listening server as specified in the LANSync Preferences panel on your Palm\&.

.SH "PILOT-NREDIR COMMAND"

.PP
\-n, \-\-net

.PP
Redirect to the network port set in the LANSync Preferences panel on the Palm\&.

.SH "OPTIONS"

                    \fB\-n\fR, \fB\-\-net\fR
                
.PP
Redirect to net:

.SS "Conduit Options"

                        \fB\-p\fR, \fB\-\-port\fR \fIport\fR
                    
.PP
Use device file \fIport\fR to communicate with the Palm handheld\&. If this is not specified, \fIpilot\-nredir \fR will look for the \fI$PILOTPORT\fR environment variable\&. If neither are found or supplied, \fI pilot\-nredir \fR will print the usage information\&.

                        \fB\-q\fR, 
                        \fB\-\-quiet\fR
                    
.PP
Suppress 'Hit HotSync button' message

                        \fB\-v\fR, \fB\-\-version\fR
                    
.PP
Display version of \fIpilot\-nredir\fR and exit without connecting\&.

.SS "Help Options"

                        \fB\-h\fR, \fB\-\-help\fR
                    
.PP
Display the help synopsis for \fIpilot\-nredir \fR and exit without connecting\&.

                        \fB\-\-usage\fR 
                    
.PP
Display a brief usage message and exit without connecting\&.

.SH "EXAMPLES"

.PP
 \fIpilot\-nredir\fR  \fB\-n\fR  \fB\-p\fR  \fI/dev/pilot\fR 

.SH "AUTHOR"

.PP
This manual page was written by Neil Williams <linux@codehelp\&.co\&.uk> 

.SH "REPORTING BUGS"

.PP
We have an online bug tracker\&. Using this is the only way to ensure that your bugs are recorded and that we can track them until they are resolved or closed\&. Reporting bugs via email, while easy, is not very useful in terms of accountability\&. Please point your browser to http://bugs\&.pilot\-link\&.org: \fIhttp://bugs.pilot-link.org\fR and report your bugs and issues there\&.

.SH "SUPPORT"

.PP
pilot\-link Development List <pilot\-link\-devel@pilot\-link\&.org> 

.SH "COPYRIGHT"

.PP
This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version\&.

.PP
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&. See the GNU General Public License for more details\&.

.PP
You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc\&., 51 Franklin St, Fifth Floor, Boston, MA 02110\-1301, USA\&.

.SH "SEE ALSO"

.PP
 \fIpilot\-link\fR(7)



src/pilot-nredir.c
/*
 * $Id: pilot-nredir.c,v 1.54 2006/10/12 14:21:21 desrod Exp $ 
 *
 * pilot-nredir.c: Redirect a connection over the network
 *
 * Copyright (C) 1997, Kenneth Albanowski
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 *
 */

#include <stdio.h>

#include "pi-dlp.h"
#include "pi-header.h"
#include "pi-userland.h"

int main(int argc, const char *argv[])
{
	int 	c,		/* switch */
		len,
		sd 		= -1,
		netsd 		= -1, /* This is the network socket */
		state;

	size_t	size;

	const char
                *progname = "pi-nredir";

	char
		port2[255] = "net:";

	pi_buffer_t *buffer;

	struct 	NetSyncInfo 	Net;

	poptContext po;

	enum { mode_none=0, mode_net=257 }
		run_mode = mode_none;

	struct poptOption options[] = {
		USERLAND_RESERVED_OPTIONS
		{"net", 'n', POPT_ARG_NONE, NULL, mode_net, "Redirect to net:", NULL},
	    POPT_AUTOHELP    
		POPT_TABLEEND
	};

	po = poptGetContext(progname, argc, argv, options, 0);
	poptSetOtherOptionHelp(po,"\n\n"
	"   Accept connection and redirect via Network Hotsync Protocol.\n"
	"   This will bind your locally connected device to a network port, and\n"
	"   redirect them through the network device to a listening server as\n"
	"   specified in the LANSync Preferences panel on your Palm.\n\n"
	"   Example arguments:\n"
	"      -n -p /dev/pilot\n\n");

	if (argc < 2) {
		poptPrintUsage(po,stderr,1);
		return 1;
	}
	while ((c = poptGetNextOpt(po)) >= 0) {
		switch(c) {
		case mode_net :
			if (mode_none == run_mode) {
				run_mode = c;
			} else {
				if (c != run_mode) {
					fprintf(stderr,"   ERROR: Specify only one mode (-n)\n");
					return 1;
				}
			}
			break;
		default:
			fprintf(stderr,"   ERROR: Unhandled option %d.\n",c);
			return 1;
		}
	}
	if (c < -1) {
		plu_badoption(po,c);
	}
	sd = plu_connect();
	if (sd < 0)
		goto error;

	if (dlp_ReadNetSyncInfo(sd, &Net) < 0) {
		fprintf(stderr,
			"   ERROR: Unable to read network information, cancelling sync.\n");
		goto error_close;
	}

	if (!Net.lanSync) {
		fprintf(stderr,
			"   ERROR: LANSync not enabled on your Palm, cancelling sync.\n");
		goto error_close;
	}

	netsd = pi_socket(PI_AF_PILOT, PI_SOCK_STREAM, PI_PF_NET);
	if (netsd < 0)
		goto error_close;

	strncat(port2, Net.hostAddress, 251 - strlen(Net.hostAddress));

	if (!plu_quiet) {
		printf("\tTrying %s... ", Net.hostAddress);
		fflush(stdout);
	}
	if (pi_connect(netsd, port2) < 0) {
		fprintf(stderr,"   ERROR: Failed to connect to network.\n");
		goto error_close;
	}

	if (!plu_quiet) {
		printf("Connected\n");
		fflush(stdout);
	}

	buffer = pi_buffer_new (0xffff);

	while ((len = pi_read(netsd, buffer, 0xffff)) > 0) {
		pi_write(sd, buffer->data, len);
		buffer->used = 0;
		len = pi_read(sd, buffer, 0xffff);
		if (len < 0)
			break;
		pi_write(netsd, buffer->data, len);
		buffer->used = 0;
	}

	pi_buffer_free (buffer);

	state = PI_SOCK_CONN_END;
	size = sizeof (state);
	pi_setsockopt (sd, PI_LEVEL_SOCK, PI_SOCK_STATE, &state, &size);
	pi_setsockopt (netsd, PI_LEVEL_SOCK, PI_SOCK_STATE, &state, &size);

	pi_close(sd);
	pi_close(netsd);

	return 0;

 error_close:
 	if (netsd >=0) {
		pi_close(netsd);
	}
	pi_close(sd);

 error:
	return -1;
}

/* vi: set ts=8 sw=4 sts=4 noexpandtab: cin */
/* ex: set tabstop=4 expandtab: */
/* Local Variables: */
/* indent-tabs-mode: t */
/* c-basic-offset: 8 */
/* End: */
